### ISSUE TITLE
syntax checking, "warning: missing terminating ' character"

### ISSUE THREAD

**Issue by ssp3nc3r** [stan-dev/rstan#793]
Checking a Stan file, either from RStudio or the console, _e.g._, 

`> rstan:::rstudio_stanc("test.stan")`

or 

`rstan:::stanc("test.stan")`

 that uses the transpose operator, *e.g.*,

    transformed data {
      vector[2] a = [1,1]';
    }

throws a warning:

```
/var/folders/pg/qc4jx60s2ys3smt604547ckw0000gn/T//RtmpJ1AeUQ/file7cf7321b3a6a.stan:2:22: warning: missing terminating ' character [-Winvalid-pp-token]
  vector[2] a = [1,1]';
                     ^
1 warning generated.
```

I suspect it's flagging the transpose as a string or something.

Rstan: 2.21.2
StanHeaders: 2.21.0-5
RStudio: 1.3.1056
R: 4.02

**Comment by pboesu** [comment]
I get the same warning message for a model with a transposed vector using 

Rstan 2.21.1
StanHeaders_2.21.0-5
R version 3.6.1 (2019-07-05)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.6

Looks to be the same issue as https://github.com/stan-dev/rstan/issues/797

**Comment by jgabry** [comment]
Yeah looks like the same as #797. Definitely a bug but I'm not sure if it's from the R code or from Stan itself. @bgoodri Do you know if this is a bug in RStan or in Stan? It looks like it's not coming from R, but I don't get this warning from CmdStanR. So that makes me think it's either a bug in RStan or a bug in Stan that was fixed in a later version than what RStan is using.

**Comment by bgoodri** [comment]
It isn't a bug. The `stanc_beta` function calls the C preprocessor to
resoloves the #includes before it tries to parse the file with the
Javascript parser. The C preprocessor applies C syntax checking rules and
under C, that is an unterminated 'string'. People can either ignore the
warning, or put

CPPFLAGS += -Wnoinvalid-pp-token

in their ~/.R/Makevars or use transpose() instead of ' to transpose a
vector. I think the latter is actually the best option because the ' is
very difficult to read anyway.

**Comment by jgabry** [comment]
Ok I guess there's some confusion since @bob-carpenter said it was definitely a bug in #797. From what @bgoodri says, it sounds like it's not _technically_ a bug, although it's pretty bug-like in that users can't use valid Stan syntax without a spurious warning. 

> CPPFLAGS += -Wnoinvalid-pp-token

No user is going to know to do this. Since rstan is now editing Makevars files this should probably get set automatically by rstan or we're going to keep getting these reports (this isn't the first time this has been reported).

**Comment by alyst** [comment]
The question is whether this syntax would be allowed in future Stan versions.
If it would be, then the current behavior is a bug.
IMHO *transpose()* is too long, so, given that you have to transpose even when declaring vector expression (`[....]'`), the Stan code would look overly tedious.

**Comment by jgabry** [comment]
> The question is whether this syntax would be allowed in future Stan versions.

As far as I know it will continue to be allowed. If there's no deprecation warning then there probably aren't any plans to remove it, although I suppose that could change between now and the next _major_ release where we allow breaking changes. 

> If it would be, then the current behavior is a bug.

Yeah thinking about it more, I think this a bug, not just bug-like (which was my previous assessment). Even though @bgoodri's explanation makes sense, if there are warnings when using valid syntax then that's a bug from the user's perspective even if it's explainable from a developer's perspective, especially since the behavior is undocumented.

**Comment by bgoodri** [comment]
It is valid Stan syntax but invalid C syntax so I'm not surprised the
preprocessor complains. Just add -Wno-... to CPPFLAGS.

**Comment by jgabry** [comment]
> Just add -Wno-... to CPPFLAGS.

That's a fine temporary solution, but I think we need to either (a) recommend everyone set this flag in the installation instructions or (b) set it automatically for the user. I'm ok either way if you have a preference, but I don't like the user having to open an issue on GitHub to find out they need to set an obscure flag just to use valid syntax without warnings.

**Comment by bgoodri** [comment]
CRAN hasn't historically allowed packages to turn off compiler warnings, so
it is something users would have to do for themselves. But I think they
should be using the transpose() function.

**Comment by jgabry** [comment]
> But I think they should be using the transpose() function.

But they have no way to know that. It's not like the warning message points them to the transpose() function. If you feel strongly about it you could propose that Stan deprecates `'` or you could edit the Stan manual to always use `transpose()` in examples so it catches on, but otherwise nobody knows they should use `transpose()` unless they happen to read this discussion deep in a GitHub issue, which we certainly can't expect.

**Comment by mark-andrews** [comment]
This warning also arises routinely for `brms` users (that's where I first encountered it). There, `brm` writes the Stan code, so using `transpose()` instead of `'` is not an option for the users. Perhaps @paul-buerkner has some opinions on what `brms` users should do, or about whether `brm` will write Stan code using `transpose()` instead of `'`. It does not yet look like a issue on this topic has been raised on https://github.com/paul-buerkner/brms.

**Comment by paul-buerkner** [comment]
Yeah i have seen this issue and then forgot about it. Will change ' to
transpose in brms.

Mark Andrews <notifications@github.com> schrieb am Sa., 5. Sept. 2020,
15:09:

**Comment by paul-buerkner** [comment]
Change brms Stan code generation to `transpose()` on github.
